### PR TITLE
learners memoize training by default

### PR DESCRIPTION
**Pull request by jeremyrcoyle**
Used the `memoise` package (https://github.com/r-lib/memoise) with a custom digest function to memoise learner fits. Partially addresses https://github.com/jeremyrcoyle/sl3/issues/9

Summary of changes: 
* added `sl3.memoise.learner` option that controls default memoise behavior (default is `TRUE` -- do memoise)
* added `memoise_learner` option to `Lrnr_base` (defaults to `sl3.memoise.learner`)
* made sure `learner$initialize` passes its dot arguments to `Lrnr_base$initialize` for all existing learners (`learner$params` also gets dot arguments)
* changed `Lrnr_base$params` from a public member to an active member pointing at a private member (prevents learner parameter mutation because it brakes memoisation assumptions)
* deleted stray `Untitled.R` file

Still to do:
- [ ] verify that we're not breaking the `memoise` api by hacking in a custom digest function
- [ ] add support for filessytem based caching
- [ ] investigate interaction between this and multisession parallelization

### CODE REVIEW

**Comment by nhejazi** [comment]
Looks good so far! (As of b239c4f.) Probably best to hold off on merging until we've covered the remaining items listed in the todo.

File: R/Lrnr_base.R [hunk 1]
```diff
@@ -19,28 +19,39 @@
...
+                       if(is.null(params)){
+                         params=list(...)
                        }
-
-                       self$params=params
+                       
+                       private$.params=params
                        private$.learner_uuid = UUIDgenerate(use.time=T)
                        
+                       if(memoise_learner){
```
Comment (osofr) [inline R/Lrnr_base.R:37]: OK, I see, pretty cool. So if memoisation is turned on, the private$.train() method is essentially placed into the private$.memoised_train?

**Review by nhejazi — APPROVED** [review]
lgtm

File: R/Lrnr_base.R (hunk 1 above)
Comment (osofr) [reply R/Lrnr_base.R:37]: I am still not following though in how does the `private$.learner_uuid` get used, in combination with `task$uuid`?

Comment (jeremyrcoyle) [reply R/Lrnr_base.R:37]: yeah I moved the old `self$train` code into `self$default_train` and now `self$train` calls that or it's memoized version depending on if memoisation is turned on

**Comment by jeremyrcoyle** [comment]
I'm disabling default memoization for now. I'll merge this change, and we can flip it back on as the default after we determine how it integrates with parallelism

File: R/Lrnr_base.R (hunk 1 above)
Comment (osofr) [reply R/Lrnr_base.R:37]: So memoisation happens within each learner instance. So the learner basically knows when it has been trained on the same or new data. That knowledge is based on uuid of the task? If I am getting a task with a uuid that has already been memoised, then I will just return the previous value. What happens if I get a task with new uuid? Does it get retrained and I am saving only the last fit object or am I saving all fit objects?

Comment (jeremyrcoyle) [reply R/Lrnr_base.R:37]: private$.learner_uuid is actually not used at all here, you get a unique cache for each learner instance by calling memoise in the constructor for the learner. See the `memA2` example here: https://www.rdocumentation.org/packages/memoise/versions/1.1.0/topics/memoise

**Review by osofr — APPROVED** [review]
Looks great!

It would be really great to have some benchmarks on how this actually speeds up the evaluation (vs. no memoising) and that it doesn't completely blow up in memory when we have a lot of objects and learners going around...

**Comment by jeremyrcoyle** [comment]
Clearly I need to write some docs on this feature. When `learner$train` is called for a memoized learner, the learner fit object is being saved in the cache (hopefully by reference). The cache is keyed (more or less) based on `task$uuid`, so a new task with a different `uuid` will generate a new entry. (See this test: https://github.com/jeremyrcoyle/sl3/blob/d203835b43d7670b2ed5259bde066ae39f416080/tests/testthat/test_memoise.R#L30) Both entires will be maintained. I definitely agree that we need to add some benchmarking to verify speedup and reasonable memory usage.

**Comment by osofr** [comment]
I am just moving my questions here, for the discussion.

So each call to learner$train(task) is memoised / saved. Then what is the behavior when calling the same `learner$train(task_1)` and then `learner$train(task_2)`? We are saving all of the fit_objects? This might end up going pretty badly. It might help to allow some kind of cache cleaning to prevent a blowup, if indeed all fit objects are now being saved. Perhaps we can have some kind of a rule, that only the last 5/10 train objects are saved and that's it?

So we are essentially changing the interface of the learner, correct? Previously, we were cloning the learner and returned self on `learner$train`:

https://github.com/jeremyrcoyle/sl3/blob/91b9190370bf6f34b7ae409b30d80a844c3aefce/R/Lrnr_base.R#L40-L51

Now we are just adding the fit object to the memoise cache and returning the exact same learner object (self), without any cloning of itself? This is interesting, but its possibly a breaking change (obviously). It would be nice to completely understand the implications of that.

Adding the update code snippet for completeness:
https://github.com/jeremyrcoyle/sl3/blob/74d7dd37c01f9912810ae6f7672ccc6b07aed599/R/Lrnr_base.R#L29-L62

**Comment by jeremyrcoyle** [comment]
Yes, we are saving all the fit objects. Under normal usage of the package the fit objects would be usually be explicitly saved anyway, as they are in all current examples/tests. For example:
https://github.com/jeremyrcoyle/sl3/blob/master/tests/testthat/test_learners_reg.R#L19

The one thing I'm not sure about is what will happen in the current implementation if we explicitly call a fit object's destructor (needs testing). My guess is that you would delete the object and free the memory, and then if you asked a memoized learner to refit to the same task it would return a broken reference and break.

No, we have not changed the interface to `Lrnr_base$train()`. If memoization is disabled, it uses `default_train`, which is exactly the old version of `Lrnr_base$train`. If memoization is enabled, it uses the memoized version of `default_train`, which simple avoids re-invoking it with arguments it's already seen. The memoization should be transparent.

**Comment by osofr** [comment]
Perfect thanks! A very nifty solution!

I agree that calling the fit object destroyer and then trying to train the memoised learner again is something that we really need to try and experiment with. It would also allow us to test that the caching is actually occurring for the R6 object pointers (cloned self) and no copying is actually being done. If these objects are getting copied then its really a lost cause...

So, from what I understand, there is absolutely no way for us to manage that cache? There is absolutely no way for us to clean the cache other than last 10 calls to $train?

**Comment by jeremyrcoyle** [comment]
It would be fairly easy to manage cache size in the way you're describing. We'd simply have to implement a new cache that checked cache size and deleted old members by rewriting the set method (e.g. here: https://github.com/r-lib/memoise/blob/master/R/cache_memory.R#L14). 

However, I don't think this would have the effect you're looking for. I think in a lot of cases it's unlikely that any one _learner instance_ gets fit on more than 10 or 11 different tasks (maybe one full task and ten training tasks), so at most you'd be saving maybe one learner fit. However, aggregated over all the learners this could still end up eating up a ton of memory. In order to manage the total memory being used by the cache, we'd have to implement a more global caching solution, rather than separate caches for each learner instance.

**Pull request merged**